app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt

```diff
@@ -0,0 +1,162 @@
+import com.epfl.drawyourpath.path.Path
+import com.epfl.drawyourpath.path.Run
+import com.epfl.drawyourpath.userProfile.dailygoal.DailyGoal
+import com.epfl.drawyourpath.utils.Utils
+import com.google.android.gms.maps.model.LatLng
+import com.google.firebase.database.DataSnapshot
+import java.time.LocalDate
+
+object FirebaseDatabaseUtils {
+
+    private fun getNumber(snapshot: DataSnapshot?): Number? {
+        return snapshot?.value as Number?
+    }
+
+    /**
+     * Helper function to obtain the daily goal list from the database of the user
+     * @param data the data snapshot containing the daily goal list
+     * @return a list containing the daily goal realized by the user
+     */
+    fun transformDailyGoals(data: DataSnapshot?): List<DailyGoal> {
+        return data?.children?.mapNotNull {
+            if (it.key == null) {
+                Log.w(this::class.java.name, "Daily goal's key was null.")
+                null
+            } else {
+                val date = LocalDate.ofEpochDay(it.key!!.toLong())
+
+                val expectedDistance: Double? =
+                    getNumber(it.child(FirebaseKeys.GOAL_HISTORY_EXPECTED_DISTANCE))?.toDouble()
+                val expectedTime: Double? =
+                    getNumber(it.child(FirebaseKeys.GOAL_HISTORY_EXPECTED_TIME))?.toDouble()
+                val expectedPaths: Int? =
+                    getNumber(it.child(FirebaseKeys.GOAL_HISTORY_EXPECTED_PATHS))?.toInt()
+                val distance: Double? =
+                    getNumber(it.child(FirebaseKeys.GOAL_HISTORY_DISTANCE))?.toDouble()
+                val time: Double? =
+                    getNumber(it.child(FirebaseKeys.GOAL_HISTORY_TIME))?.toDouble()
+                val paths: Int? =
+                    getNumber(it.child(FirebaseKeys.GOAL_HISTORY_PATHS))?.toInt()
+
+                DailyGoal(
+                    expectedDistance = expectedDistance ?: 0.0,
+                    expectedPaths = expectedPaths ?: 0,
+                    expectedTime = expectedTime ?: 0.0,
+                    distance = distance ?: 0.0,
+                    time = time ?: 0.0,
+                    paths = paths ?: 0,
+                    date = date,
+                )
+            }
+        } ?: emptyList()
+    }
+
+    /**
+     * Helper function to obtain a list from the keys of a database snapshot.
+     * @param data the data snapshot to be converted to a list
+     * @return a list containing the keys of the snapshot
+     */
+    fun getKeys(data: DataSnapshot?): List<String> {
+        return data?.children?.mapNotNull { it.key as String } ?: emptyList()
+    }
+
+    /**
+     * Helper function to obtain the runs history from the database of the user
+     * @param data the data snapshot containing the history
+     * @return a list containing the history of the runs of the user
+     */
+    fun transformRuns(data: DataSnapshot?): List<Run> {
+        return data?.children?.mapNotNull {
```

Comment:
The name of this function is a little bit ambiguous with transformRun . Maybe you can rename it : transformRunList(...)